### PR TITLE
LS-Restrict algorithm for Group Closeness 

### PR DESCRIPTION
**Pull request by angriman**
This PR brings the LS-restrict and the Grow-Shrink algorithms presented in the paper "Local Search for Group Closeness Maximization on Big Graphs" presented at IEEE BigData 2019.

### CODE REVIEW

**Comment by angriman** [comment]
The new algorithms also support AVX instructions, which are used if the `NETWORKIT_NATIVE` flag is set to `ON`, and if they are available on the machine. This operation can be done easily on Linux and MacOS, but I could not find a solution on Windows. Does anyone know how to do this on Windows?

File: CMakeLists.txt [hunk 1]
```diff
@@ -95,6 +95,54 @@ else()
...
+		set(AVX_512vl_TRUE ${AVX_FOUND} PARENT_SCOPE)
+	endfunction(check_avx512vl)
+
+	# Check if AVX instructions are supported on host. If yes, enable the required AVX flags.
+	function(set_avx_flags CPUINFO)
+		check_avx512f(${CPUINFO})
+		check_avx512vl(${CPUINFO})
+
+		if (AVX_512f_TRUE AND AVX_512vl_TRUE)
+			set(NETWORKIT_CXX_FLAGS "${NETWORKIT_CXX_FLAGS} -mavx -mavx512vl -mavx512f -DNETWORKIT_WITH_AVX")
```
Comment (fabratu) [inline CMakeLists.txt:124]: I wonder if the support for `AVX512` necessary here. You use 256bit-width variables (`__m256i`), which are also supported by `AVX2` ([Intrinsics Guide from Intel](__m256i)). This would lead to many more users/platforms able to use it. Does 512 vs 2 result in a speed-up?

Disclaimer:
Not an expert on this subject, but the common sense (at least on Skylake microarchitecture) is, that a speed-up in real-life applications is only expected if 512-bit width variables are used. And even then it is sublinear, since `AVX512` functions have a higher latency.

File: include/networkit/centrality/GroupClosenessLocalSwaps.hpp [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+                return false;
+            }
+        }
+        return true;
+    }
+
+    typedef struct {
+        uint64_t state;
+        uint64_t inc;
+    } pcg32_random_t;
```
Comment (fabratu) [inline include/networkit/centrality/GroupClosenessLocalSwaps.hpp:152]: Based on the discussion in PR #415, is PCG still needed?

Comment (angriman) [reply include/networkit/centrality/GroupClosenessLocalSwaps.hpp:152]: No, I will remove it.

File: include/networkit/centrality/GroupClosenessGrowShrink.hpp [hunk 3]
```diff
@@ -0,0 +1,665 @@
+/*
```
Comment (fabratu) [inline include/networkit/centrality/GroupClosenessGrowShrink.hpp:1]: Whats the rationale behind implementing `GrowClosenessGrowShrink` and `Dijkstra` as header only?

File: include/networkit/centrality/GroupClosenessGrowShrink.hpp [hunk 4]
```diff
@@ -0,0 +1,665 @@
...
+            throw std::runtime_error("Error, empty group.");
+        }
+    }
+
+    // Easier constructor to wrap in Cython.
+    GroupClosenessGrowShrinkGeneral(const Graph &graph, const std::vector<node> &group,
+                                    bool extended = false, count insertions = 0)
+        : GroupClosenessGrowShrinkGeneral(graph, group.begin(), group.end(), extended, insertions) {
+    }
+
```
Comment (fabratu) [inline include/networkit/centrality/GroupClosenessGrowShrink.hpp:67]: Please add a default-destrctor.

File: include/networkit/centrality/GroupClosenessLocalSwaps.hpp [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+        }
+        if (G->isWeighted()) {
+            WARN("This algorithm does not support edge Weights, they will be ignored.");
+        }
+    }
+
+    // Easier constructor to wrap in Cython.
+    GroupClosenessLocalSwaps(const Graph &graph, const std::vector<node> &group)
+        : GroupClosenessLocalSwaps(graph, group.begin(), group.end()) {}
+
```
Comment (fabratu) [inline include/networkit/centrality/GroupClosenessLocalSwaps.hpp:54]: Please add a default-destructor.

File: include/networkit/centrality/GroupClosenessGrowShrink.hpp [hunk 6]
```diff
@@ -0,0 +1,665 @@
...
+            std::fill(visited.begin(), visited.end(), 0);
+        }
+    }
+
+    typedef struct {
+        uint64_t state;
+        uint64_t inc;
+    } pcg32_random_t;
+
+    std::vector<pcg32_random_t> rng;
```
Comment (fabratu) [inline include/networkit/centrality/GroupClosenessGrowShrink.hpp:195]: Same question about PCG here.

File: include/networkit/centrality/GroupClosenessLocalSwaps.hpp (hunk 5 above)
Comment (avdgrinten) [reply include/networkit/centrality/GroupClosenessLocalSwaps.hpp:54]: Why?

File: include/networkit/centrality/GroupClosenessGrowShrink.hpp (hunk 3 above)
Comment (angriman) [reply include/networkit/centrality/GroupClosenessGrowShrink.hpp:1]: Because they are templated, moving the implementation to a source file would yield compile-time errors.

Comment (avdgrinten) [reply include/networkit/centrality/GroupClosenessGrowShrink.hpp:1]: A templated Dijskstra is reasonable but why do we need a templated Grow-Shrink?

File: include/networkit/centrality/GroupClosenessGrowShrink.hpp (hunk 4 above)
Comment (avdgrinten) [reply include/networkit/centrality/GroupClosenessGrowShrink.hpp:67]: Why?

File: include/networkit/centrality/GroupClosenessGrowShrink.hpp (hunk 3 above)
Comment (angriman) [reply include/networkit/centrality/GroupClosenessGrowShrink.hpp:1]: To handle weighted/unweighted graphs. I wanted to specialize the code for the two version to avoid doing many run-time checks for `G.isWeighted()` .

File: include/networkit/centrality/GroupClosenessGrowShrink.hpp [hunk 7]
```diff
@@ -0,0 +1,665 @@
...
+     * Runs the algorithm.
+     */
+    void run() override {
+        init();
+
+        if (!insertions) {
+            Diameter diam(*G, DiameterAlgo::estimatedRange, 0.1);
+            diam.run();
+            DEBUG("Diameter upper bound = ", diam.getDiameter().second);
+            insertions = std::max(1., .5
```
Comment (fabratu) [inline include/networkit/centrality/GroupClosenessGrowShrink.hpp:106]: Just to be sure: Is the truncation from `double` to `int` intended?

Comment (angriman) [reply include/networkit/centrality/GroupClosenessGrowShrink.hpp:106]: Yes, as we want the number of insertions to be an integer >= 1.

File: include/networkit/centrality/GroupClosenessGrowShrink.hpp (hunk 3 above)
Comment (avdgrinten) [reply include/networkit/centrality/GroupClosenessGrowShrink.hpp:1]: As far as I can see, this does not remove the need to perform runtime checks, it only specializes the code for `count` weights instead of `double` weights. I'm not sure if that is really worthwhile (double math should only be slightly slower than integer math). If you deem it worthwhile, use the pimpl idiom to avoid putting the entire algorithm into the header.

File: include/networkit/centrality/GroupClosenessLocalSwaps.hpp [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+        uint16_t items[16];
+        __m256i vec;
+    };
+    std::vector<RandItem> randVec;
+#else
+    std::vector<uint16_t> randVec;
+#endif // NETWORKIT_WITH_AVX
+
+    void resetGamma(node x, index idx) {
+        std::fill(gamma.begin() + group.size() * x, gamma.begin() + group.size() * (x + 1), 0);
```
Comment (fabratu) [inline include/networkit/centrality/GroupClosenessLocalSwaps.hpp:123]: Is it safe, that `gamma.begin() + group.size() * (x + 1)` will not go oob?

Comment (angriman) [reply include/networkit/centrality/GroupClosenessLocalSwaps.hpp:123]: Yes because `gamma` has size `n * group.size()`, and `x` is always at most `n - 1`.

**Comment by angriman** [comment]
@fabratu I updated the avx2 flags, now this can be reviewed.

File: include/networkit/generators/quadtree/QuadNodeCartesianEuclid.hpp [hunk 9]
```diff
@@ -391,7 +391,6 @@ class QuadNodeCartesianEuclid final {
                     if (i >= lsize) break;
                     TRACE("Jumped with delta ", delta, " arrived at ", i);
                 }
-                assert(i >= 0);
```
Comment (avdgrinten) [inline include/networkit/generators/quadtree/QuadNodeCartesianEuclid.hpp:394]: Unrelated change; please remove.

File: include/networkit/generators/quadtree/QuadNodePolarEuclid.hpp [hunk 10]
```diff
@@ -503,7 +503,6 @@ class QuadNodePolarEuclid final {
                     if (i >= lsize) break;
                     TRACE("Jumped with delta ", delta, " arrived at ", i);
                 }
-                assert(i >= 0);
```
Comment (avdgrinten) [inline include/networkit/generators/quadtree/QuadNodePolarEuclid.hpp:506]: Same.

File: networkit/cpp/centrality/GroupClosenessLocalSwaps.cpp [hunk 11]
```diff
@@ -0,0 +1,362 @@
...
+
+    stack.assign(n, 0);
+    farness.assign(group.size(), 0);
+    farnessDecrement.assign(group.size(), 0);
+    sumOfMins.assign(n, 0);
+
+    intDistributions.resize(omp_get_max_threads());
+    urngs.reserve(omp_get_max_threads());
+    for (index i = 0; i < static_cast<index>(omp_get_max_threads()); ++i)
+        urngs.emplace_back(Aux::Random::getURNG());
```
Comment (avdgrinten) [inline networkit/cpp/centrality/GroupClosenessLocalSwaps.cpp:36]: This looks broken since the loop is not parallel and we get several identical references to the URNG of thread 0.

Comment (avdgrinten) [reply networkit/cpp/centrality/GroupClosenessLocalSwaps.cpp:36]: I also wonder if this is necessary at all. `urngs` is only used in a `parallelForNodes` loop in `initRandomVector`. Accessing the thread-local variable is cheap compared to the loop. Why don't we just do the following in `initRandomVector`:
```cpp
#pragma omp parallel
{
    auto &urng = Aux::Random::getURNG();

    #pragma omp for
    for(node u = 0; u < G.upperNodeIdBound(); ++u) {
        [...]
    }
}

File: networkit/cpp/centrality/GroupClosenessLocalSwaps.cpp [hunk 12]
```diff
@@ -0,0 +1,362 @@
...
+
+    totalSwaps = 0;
+    hasRun = false;
+}
+
+void GroupClosenessLocalSwaps::run() {
+    init();
+
+    while (findAndSwap() && ++totalSwaps < maxSwaps) {
+    }
```
Comment (avdgrinten) [inline networkit/cpp/centrality/GroupClosenessLocalSwaps.cpp:58]: This loops strange. IMHO the following is cleaner:
```suggestion
    while (findAndSwap() && ++totalSwaps < maxSwaps)
        ; // Just iterate.
```

File: networkit/cpp/centrality/GroupClosenessLocalSwaps.cpp [hunk 13]
```diff
@@ -0,0 +1,362 @@
...
+    for (const auto &idx : idxMap) {
+        if (canSwap[idx.second]) {
+            int64_t curImprovement =
+                farnessDecrementV - farness[idx.second] + farnessDecrement[idx.second];
+            if (curImprovement > improvement) {
+                improvement = curImprovement;
+                u = idx.first;
+            }
+        }
+    }
```
Comment (avdgrinten) [inline networkit/cpp/centrality/GroupClosenessLocalSwaps.cpp:91]: Why do we need this second loop? What does it do? Maybe add a comment.

File: networkit/cpp/centrality/GroupClosenessLocalSwaps.cpp [hunk 14]
```diff
@@ -0,0 +1,362 @@
...
+    gamma.assign(n * group.size(), false);
+    canSwap.assign(group.size(), false);
+    visited.assign(n, false);
+
+    idxMap.clear();
+    idxMap.reserve(group.size());
+
+    stack.assign(n, 0);
+    farness.assign(group.size(), 0);
+    farnessDecrement.assign(group.size(), 0);
```
Comment (avdgrinten) [inline networkit/cpp/centrality/GroupClosenessLocalSwaps.cpp:30]: Decrement -> Decrease. Similarly, if that occurs elsewhere: Increment -> Increase.

File: networkit/cpp/centrality/GroupClosenessLocalSwaps.cpp [hunk 15]
```diff
@@ -0,0 +1,362 @@
...
+
+int64_t GroupClosenessLocalSwaps::computeFarnessDecrement(node v) {
+    std::fill(visited.begin(), visited.end(), false);
+
+    std::queue<node> q;
+    q.push(v);
+    distance[v] = 0;
+    visited[v] = true;
+
+    int16_t decrement{1};
```
Comment (avdgrinten) [inline networkit/cpp/centrality/GroupClosenessLocalSwaps.cpp:201]: uint16_t seems to be a mistake here (?).

File: include/networkit/centrality/GroupClosenessLocalSwaps.hpp [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+    void init();
+    void bfsFromGroup();
+    bool findAndSwap();
+    node estimateHighestDecrement();
+    void initRandomVector();
+    int64_t computeFarnessDecrement(node u);
+    void resetGamma(node x, index idx);
+
+    // Returns true if the vertex at the give index is the only nearest vertex to x
+    bool gammaIsU(node x, index idx) const;
```
Comment (avdgrinten) [inline include/networkit/centrality/GroupClosenessLocalSwaps.hpp:97]: Why is that needed? Seems that this is not called.

**Review by avdgrinten — CHANGES_REQUESTED** [review]

File: networkit/cpp/centrality/GroupClosenessLocalSwaps.cpp (hunk 13 above)
Comment (angriman) [reply networkit/cpp/centrality/GroupClosenessLocalSwaps.cpp]: That's not needed, I probably forgot to remove it from a previous version. I removed it together with the `canSwap` vector, which was used just here.

File: include/networkit/centrality/GroupClosenessLocalSwaps.hpp (hunk 16 above)
Comment (angriman) [reply include/networkit/centrality/GroupClosenessLocalSwaps.hpp]: True, I removed it.

File: networkit/cpp/centrality/GroupClosenessLocalSwaps.cpp (hunk 15 above)
Comment (angriman) [reply networkit/cpp/centrality/GroupClosenessLocalSwaps.cpp]: Yes, that should be `int64_t`.

**Review by avdgrinten — APPROVED** [review]
LGTM now.
